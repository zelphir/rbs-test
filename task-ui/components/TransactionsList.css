table {
  margin-top: 60px;
  border-collapse: collapse;
  width: 100%;
}

th {
  position: sticky;
  top: 60px;
  color: #d5dde5;
  border-right: 1px solid #c1c3d1;
  background: #333;
  font-size: 16px;
  font-weight: 100;
  padding: 0 5px;
  height: 50px;
  text-align: left;
  vertical-align: middle;
}

th:last-child {
  border-right: none;
}

tr {
  border-top: 1px solid #c1c3d1;
  border-bottom: 1px solid #c1c3d1;
  color: #666b85;
  font-size: 16px;
  font-weight: normal;
  text-shadow: 0 1px 1px rgba(256, 256, 256, 0.1);
}

tr:hover td {
  background: #4e5066;
  color: #ffffff;
  border-top: 1px solid #22262e;
}

tr:first-child {
  border-top: none;
}

tr:last-child {
  border-bottom: none;
}

tr:nth-child(odd) td {
  background: #ebebeb;
}

tr:nth-child(odd):hover td {
  background: #4e5066;
}

td {
  background: #ffffff;
  padding: 10px 5px;
  text-align: left;
  vertical-align: middle;
  font-weight: 300;
  font-size: 16px;
  border-right: 1px solid #c1c3d1;
}

td:last-child {
  border-right: 0px;
}
