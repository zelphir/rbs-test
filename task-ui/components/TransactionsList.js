import React from 'react'
import PropTypes from 'prop-types'
import { searchByName, sortByAmount } from 'task-library'
import TableRow from './TableRow'
import Toolbar from './Toolbar'
import TableHead from './TableHead'

import './TransactionsList.css'

class TransactionsList extends React.PureComponent {
  static propTypes = {
    transactions: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        amount: PropTypes.number.isRequired,
        date: PropTypes.string.isRequired,
        business: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        account: PropTypes.string.isRequired
      })
    ).isRequired
  }

  state = {
    transactions: this.props.transactions
  }

  searchBy = event => {
    const value = event.target.value

    this.setState({
      transactions: searchByName(this.props.transactions, value)
    })
  }

  sortBy = event => {
    event.preventDefault()
    const order = event.target.dataset.order

    this.setState(prevState => ({
      transactions: sortByAmount(prevState.transactions, order)
    }))
  }

  render() {
    const headers = this.props.transactions[0]
    const { transactions } = this.state

    return (
      <React.Fragment>
        <Toolbar searchBy={this.searchBy} sortBy={this.sortBy} />
        <table>
          <TableHead headers={headers} />
          <tbody>
            {transactions.map(transaction => (
              <TableRow key={transaction.id} transaction={transaction} />
            ))}
          </tbody>
        </table>
      </React.Fragment>
    )
  }
}

export default TransactionsList
