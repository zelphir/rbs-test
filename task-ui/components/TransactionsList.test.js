import React from 'react'
import { shallow } from 'enzyme'
import TransactionsList from './TransactionsList'

const transactions = [
  {
    id: 1,
    amount: 85.91,
    date: '2012-02-02T00:00:00.000Z',
    business: 'Bode - Kreiger',
    name: 'Savings Account 6670',
    type: 'withdrawal',
    account: '86389677'
  },
  {
    id: 2,
    amount: 45.91,
    date: '2012-04-02T00:00:00.000Z',
    business: 'Bode - Kreiger',
    name: 'Money Account 4582',
    type: 'deposit',
    account: '86389532'
  }
]

let wrapper

beforeEach(() => {
  wrapper = shallow(<TransactionsList transactions={transactions} />)
})

describe('<TransactionsList />', () => {
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('searchBy | sortBy', () => {
    it('searchBy should set the filtered transactions state', () => {
      expect(wrapper.state().transactions).toEqual(transactions)
      wrapper.instance().searchBy({ target: { value: 'money' } })
      expect(wrapper.state().transactions).toEqual([transactions[1]])
    })

    it('sortBy should set the ordered transactions state', () => {
      expect(wrapper.state().transactions).toEqual(transactions)
      wrapper
        .instance()
        .sortBy({ target: { dataset: { order: 'ASC' } }, preventDefault: jest.fn() })
      expect(wrapper.state().transactions).toEqual([transactions[1], transactions[0]])
    })
  })
})
