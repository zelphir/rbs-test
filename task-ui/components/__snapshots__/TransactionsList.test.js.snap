// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionsList /> should match the snapshot 1`] = `
<React.Fragment>
  <Toolbar
    searchBy={[Function]}
    sortBy={[Function]}
  />
  <table>
    <TableHeader
      headers={
        Object {
          "account": "86389677",
          "amount": 85.91,
          "business": "Bode - Kreiger",
          "date": "2012-02-02T00:00:00.000Z",
          "id": 1,
          "name": "Savings Account 6670",
          "type": "withdrawal",
        }
      }
    />
    <tbody>
      <TableRow
        key="1"
        transaction={
          Object {
            "account": "86389677",
            "amount": 85.91,
            "business": "Bode - Kreiger",
            "date": "2012-02-02T00:00:00.000Z",
            "id": 1,
            "name": "Savings Account 6670",
            "type": "withdrawal",
          }
        }
      />
      <TableRow
        key="2"
        transaction={
          Object {
            "account": "86389532",
            "amount": 45.91,
            "business": "Bode - Kreiger",
            "date": "2012-04-02T00:00:00.000Z",
            "id": 2,
            "name": "Money Account 4582",
            "type": "deposit",
          }
        }
      />
    </tbody>
  </table>
</React.Fragment>
`;
